tensorboard/webapp/plugins/npmi/safe_encoder.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""A safe method to encode numpy arrays containing nan, inf or -inf."""
+import json
+import numpy as np
+
+
+class SafeEncoder(json.JSONEncoder):
+    """Encoder that can handle numpy nan, inf, and -inf."""
+
+    def __init__(self, nan_str="null", **kwargs):
+        super(SafeEncoder, self).__init__(**kwargs)
+        self.nan_str = nan_str
+
+    def iterencode(self, o, _one_shot=False):
+        """Encode the given object and yield each string
+        representation as available.
+
+        Args:
+            o: The object to encode.
+        """
+        if self.check_circular:
+            markers = {}
+        else:
+            markers = None
+        if self.ensure_ascii:
+            _encoder = json.encoder.encode_basestring_ascii
+        else:
+            _encoder = json.encoder.encode_basestring
+
+        def floatstr(
+            o,
+            allow_nan=self.allow_nan,
+            _inf=json.encoder.INFINITY,
+            _neginf=-json.encoder.INFINITY,
+            nan_str=self.nan_str,
+        ):
+            """Makes a string out of a float value
+
+            Args:
+                o: The object to encode.
+                allow_nan: Whether nan values should be encoded or raise an
+                    Exception.
+                _inf: An infinity identifier for the encoder.
+                _neginf: A identifier for negative inf values for the encoder.
+                nan_str: The string with which to replace nan values.
+            """
+            if o != o:
+                text = nan_str
+            elif o == _inf:
+                text = "Infinity"
+            elif o == _neginf:
+                text = "-Infinity"
+            else:
+                return str(o)
+            if not allow_nan:
+                raise ValueError(
+                    "Out of range float values are not JSON compliant: "
+                    + repr(o)
+                )
+            return text
+
+        _iterencode = json.encoder._make_iterencode(
```

Comment:
Okay - my bad for not realizing that JSONEncoder only calls `default()` for types that aren't handled by the existing serialization, so I assume that's why you are taking this approach.  Sorry for not looking into that more carefully before suggesting it.

That said, it's unfortunately still not great to be calling `json.encoder._make_iterencode` since it's not part of the public API of the `json` module, meaning it could break between python releases.

Perhaps instead what we can do is take the result of the `tolist()` conversion of the tensor value and preprocess it to convert any instances of NaN to a Python `None` value, and then pass it into unmodified `json.dumps()`?  I think the conversion would be a simple as:

```
def convert(arr):
    return [convert(e) if isinstance(e, list) else None if math.isnan(e) else e for e in arr]
```

Reply:
Yes, I had to take this approach, but I agree, it is not super intuitive and can introduce its own problems.

Did what you said in [efb1128](https://github.com/tensorflow/tensorboard/pull/3799/commits/efb1128fc092fef7caac5b1055570bc1882df82a)